src/tmpfiles/tmpfiles.c

```diff
@@ -2003,6 +2005,177 @@ static int glob_item_recursively(Item *i, fdaction_t action) {
         return r;
 }
 
+static int rm_if_wrong_type_safe(
+                mode_t mode,
+                int parent_fd,
+                const struct stat *parent_st /* Only used if follow is true. */,
+                const char *name,
+                bool follow) {
+        _cleanup_free_ char *parent_name = NULL;
+        struct stat st;
+        int r;
+        const char *e;
+
+        assert(name);
+        assert((mode & ~S_IFMT) == 0);
+        if (follow)
+                assert(parent_st);
+
+        e = strrchr(name, '/');
+        if (e)
+                return -EINVAL;
+
+        r = fstatat(parent_fd, name, &st, follow ? 0 : AT_SYMLINK_NOFOLLOW);
+        if (r < 0) {
+                if (errno != ENOENT) {
+                        (void) fd_get_path(parent_fd, &parent_name);
+                        (void) log_error_errno(errno, "fstatat(%s, %s): %m", strna(parent_name), name);
+                }
+                return -errno;
+        }
+
+        /* Fail before removing anything if this is an unsafe transition. */
+        /* Ideally this would be fs-util's unsafe_transition, but it is not exported. */
+        if (follow && parent_st->st_uid != 0 && parent_st->st_uid != st.st_uid) {
+                (void) fd_get_path(parent_fd, &parent_name);
+                return log_error_errno(-ENOLINK, "unsafe transition \"%s/%s\": %m", strna(parent_name), name);
```

Comment:
pass `SYNTHETIC_ERRNO(ENOLINK)` as first arg

Reply:
to stay in line with most of our log messages start with uppercase characters, as in usual english